token/program-2022/src/processor.rs

```diff
@@ -889,6 +927,15 @@ impl Processor {
                 .base
                 .is_owned_by_system_program_or_incinerator()
             {
+                if let Ok(cpi_guard) = source_account.get_extension::<CpiGuard>() {
+                    if cpi_guard.lock_cpi.into()
+                        && in_cpi()
+                        && !cmp_pubkeys(destination_account_info.key, &source_account.base.owner)
+                    {
+                        return Err(TokenError::CpiGuardCloseAccountBlocked.into());
+                    }
```

Comment:
I think we'll need to relax this check if `source_account.base.close_authority` is not None.  Otherwise, we can get this kinda hilarious "attack":

* I create an account for you, set my PDA as close authority
* I set ownership of the account to you
* you add CPI guard
* I can never close your account through my program

Does that make sense?

Reply:
the program could still close the account, theyd just have to direct the lamports to me and not them

although it does raise the interesting question of, is that how this should work? since the original intent of close authority was to allow programs to reclaim lamports from user accounts the program funded...

the broader issue here is adversarial: cpi guard is meant to give _more_ control to users, whereas close authority takes control _away_ from users. fundamentally these two features are in tension

since rent exemption lamps are of negligible value, perhaps the check should look like "IF cpi guard enabled AND in cpi AND mint is native THEN lamports must go to owner"? i had excessive concern about close authority a few weeks ago since i hadnt learned until refactoring `command_gc` that it cant be used to steal tokens. now i kind of regard it as just... obnoxious, but with a very bad edge case for native mint

i think my ideal solution is:
* further limit this check to native mint accounts
* improve `ImmutableOwner` to also ban setting close authority (separate pr)
* hopefully someday deprecate it

Reply:
The close authority was specifically added for a project that was creating loads and loads of non-native accounts for their users, but with the ability to reclaim the lamports if they never use them. On the flipside, it's a real edge case, since there probably aren't any projects that are closing user accounts through a program.

If they really wanted to reclaim the account, they could change their program to set the authority to a hot wallet and then do it normally.  I'll retract this one too, seems like too much of an edge case, and there are ways around it if needed.